src/handlers/RF24StreamLogHandler.cpp

```diff
@@ -0,0 +1,217 @@
+}
+
+void RF24StreamLogHandler::log(uint8_t logLevel,
+                                   const __FlashStringHelper *vendorId,
+                                   const char *message,
+                                   va_list *args)
+{
+   if(logLevel > this->logLevel)
+   {
+      return;
+   }
+
+   appendTimestamp();
+   appendLogLevel(logLevel);
+   appendVendorId(vendorId);
+
+   // print formatted message
+   appendFormattedMessage(message, args);
+   stream->println("");
+}
+
+void RF24StreamLogHandler::log(uint8_t logLevel,
+                                   const __FlashStringHelper *vendorId,
+                                   const __FlashStringHelper *message,
+                                   va_list *args)
+{
+   if(logLevel > this->logLevel)
+   {
+      return;
+   }
+
+   appendTimestamp();
+   appendLogLevel(logLevel);
+   appendVendorId(vendorId);
+
+   // print formatted message
+   appendFormattedMessage(message, args);
+   stream->println("");
+}
+
+void RF24StreamLogHandler::setLogLevel(uint8_t logLevel)
+{
+   this->logLevel = logLevel;
+}
+
+void RF24StreamLogHandler::appendFormattedMessage(const char *format, va_list *args)
+{
+   for (; *format != 0; ++format)
+   {
+      if (*format == '%')
+      {
+         ++format;
+         appendFormat(*format, args);
+      }
+      else
+      {
+         stream->print(*format);
+      }
+   }
+}
+
+void RF24StreamLogHandler::appendFormattedMessage(const __FlashStringHelper *format, va_list *args)
+{
+   PGM_P p = reinterpret_cast<PGM_P>(format);
+   char c = pgm_read_byte(p++);
+   for (; c != 0; c = pgm_read_byte(p++))
+   {
+      if (c == '%')
+      {
+         c = pgm_read_byte(p++);
+         appendFormat(c, args);
+      }
+      else
+      {
+         stream->print(c);
+      }
+   }
+}
+
+void RF24StreamLogHandler::appendFormat(const char format, va_list *args)
+{
+   if (format == 's')
+   {
+      // print text from RAM
+      register char *s = (char *)va_arg(*args, int);
+      stream->print(s);
+
+      return;
+   }
+
+   if (format == 'S')
+   {
+      // print text from FLASH
+      register __FlashStringHelper *s = (__FlashStringHelper *)va_arg(*args, int);
+      stream->print(s);
+
+      return;
+   }
+
+   if (format == 'D' || format == 'F')
+   {
+      // print as double
+      stream->print(va_arg(*args, double));
+
+      return;
+   }
+
+   if (format == 'd' || format == 'i')
+   {
+      // print as integer
+      stream->print(va_arg(*args, int), DEC);
+
+      return;
+   }
+
+   stream->print(format);
+}
+
+void RF24StreamLogHandler::appendTimestamp()
+{
+   char c[12];
+   sprintf(c, "%10lu ", millis());
+   stream->print(c);
+   stream->print(" ");
+}
+
+void RF24StreamLogHandler::appendLogLevel(uint8_t logLevel)
+{
+   uint8_t logMainLevel = logLevel & 0b11111000;
+   uint8_t subLevel = logLevel & 0b00000111;
+
+   switch(logMainLevel)
+   {
+      case RF24LogLevel::ERROR:
+      {
+         stream->print(rf24LogLevels[0]);

```

Comment:
ok, but I wanted to avoid ```%10``` and ```/10```. Division by 10 is "not a natural" operation for a binary processor and it may results in more FLASH usage. It is better to have 8 (the power of 2) sublevels and use shift operator ```<<``` or ```>>```, which are piece of cake for a processor.

Reply:
no worries. I just remember a comment in RF24.cpp saying 

>switch is evil and eats ram

or something like that 😆 and then it goes into a `if ... else if ... else if ... else if ... else if ... else if ... else`

I like the shifting idea!

Reply:
I wasn't aware about those differences between ```switch``` and ``if ... else if ...```. I would like to compare the assembler listings of both solutions.

Reply:
I don't see much difference between **switch** and **if...else if...else** if it goes about **FLASH** memory usage.
Switch variant takes 0x86 bytes:
```0000097a l     F .text	00000086 _ZN20RF24StreamLogHandler14appendLogLevelEh```
while **if...else...** variant takes 0x84 bytes:
```0000097a l     F .text	00000084 _ZN20RF24StreamLogHandler14appendLogLevelEh```

I didn't analyze the RAM usage because it requires a bit deeper dissasembler code analysis.